### DISCUSSION TITLE
Is it possible to `import` other things?

### DISCUSSION THREAD

**Discussion by kgoggin** [johnlindquist/kit#742] [Q&A]
Newbie here! I'm working on tracking things in a locally running Postgres DB, and I'd love to use Script Kit to read/write from it! I'm most comfortable doing DB stuff with [Prisma](https://www.prisma.io/), so I was thinking it'd be amazing if I could `import` an instance of a Prisma client.

Looking at the `.kenv` structure, it _feels_ like I could maybe add Prisma as a dep in package.json and generate a client just like in any other project. So I tried that, and then tried two methods of working with the prisma client:
  
  1. accessing it via a new `db.ts` file, where I instantiated a prisma client and exported it as suggested in the Prisma docs, and then using an `import` in my script to pull it in. I got an error `SyntaxError: Named export 'PrismaClient' not found. The requested module '@prisma/client' is a CommonJS module...`.
  2. using `const { PrismaClient } = await npm("@prisma/client");` in the script to try pulling it in directly. TS can't tell what `PrismaClient` is here so that's a bummer, and the script errored out anyway. 

So, any ideas how I might get this working? Being able to use Prisma in scripts would be 💯 for soooo many applications! Thanks!

**Comment by johnlindquist** [comment]
Yes!

## `npm()` and TypeScript

Unfortunately, it's impossible to provide typings for `npm`*, BUT:

### Manually installing npm packages and using `await import`

<img width="889" alt="CleanShot 2022-02-25 at 13 55 32@2x" src="https://user-images.githubusercontent.com/36073/155801628-185a5935-c889-4dd0-bb3e-e2339a4e7a6a.png">

The tools in the kit npm UI will help you manage which packages are available in your `.kenv`.

After your package is installed, you can use: `await import()` and TypeScript should work just fine (unless you're using the built-in kit editor, loading typings isn't built-in yet)

`const {PrismaClient} = await import("@prisma/client")`

### Manually installing from the terminal

If those kit `npm` tools fail, you can use the traditional method:

```bash
cd ~/.kenv && ~/.kit/node/bin/npm i @prisma/client
```

\* I may try to write a vs code plugin in the future that loads in typings based on the `npm` string

> Note: I've never used Prisma, so if this isn't working for some other reason, let me know, and I'll see if I can toss a demo together

**Reply by kgoggin** [reply]
Thanks for the quick reply! Just tried out using `const { PrismaClient } = await import("@prisma/client");`, but I'm still seeing an error:

```
TypeError: PrismaClient is not a constructor
```

If I try console logging the value of PrismaClient, it logs `__no_value__` which is... weird? Gotta knock off for today but might try a few more things later if I have a chance. In the meantime if you're able to get it working I'd love to know how! Thanks so much!

**Reply by johnlindquist** [reply]
Looks like `PrimsaClient` is on the default export:

```js
// Name: Prisma Test

import "@johnlindquist/kit"
import { PrismaClient } from "@primsa/client"

// Also works, just clunky
// let { default: { PrismaClient } } = await import("@prisma/client")

let p = new PrismaClient()
```

Then I run into this error: https://stackoverflow.com/questions/67746885/prisma-client-did-not-initialize-yet-please-run-prisma-generate-and-try-to-i

You're probably more familiar with Prisma generate than I...

## __no_value__

The logger can't log `undefined` since it's not a string, so I arbitrarily chose `__no_value__` so it would stick out in the logs (it's used in other places as well). I can definitely make it more clear when the user is trying to log something that's undefined.
